docs/dinamicko-programiranje/tiling.md

```diff
@@ -0,0 +1,134 @@
+---
+title: Popločavanje
+---
+
+### Problem
+
+Zadana je ploča veličine $2 * n$ koju je potrebno popločiti pločicama dimenzija $2 * 1$. Pločice je moguće rotirati. Na koliko načina je moguće ostvariti popločavanje? Pošto broj načina može biti jako velik, ispišite rješenje modulo 1e9 + 7.
+
+### Rješenje
+
+Problem ćemo riješiti rekurzivno. Ako je ploča duljine $1$, tada možemo postaviti pločicu samo vertikalno, inače imamo dvije opcije. Možemo staviti pločicu vertikalno i time smanjiti duljinu ploče za $1$ **ili** staviti pločicu horizontalno čime smo prisiljeni staviti i drugu pločicu horizontalno u iste stupce, time smanjujemo duljinu ploče za $2$. Tada se naša relacija svodi na sljedeće $F(x) = F(x - 1) + F(x - 2)$ što prepoznajemo kao *Fibbonacijev niz*.
+
+```cpp
+const int MAXN = 1e5 + 5;
+int n;
+int mod = 1e9 + 7;
+int memo[MAXN];
+
+int solve(int ostalo){
+    if(ostalo == 0) return 1;
+    if(ostalo == 1) return 1;
+    if(memo[ostalo] != -1) return memo[ostalo];
+    return memo[ostalo] = (solve(ostalo - 1) + solve(ostalo - 2)) % mod;
+}
+
+int main(){
+    memset(memo, -1, sizeof memo);
+    cin >> n;
+    cout << solve(n);
+    return 0;
+}
+```
+Složenost ovog rješenja je $O(n)$ jer svako stanje posjećujemo samo jednom.
+
+### Teži problem
+
+Zadana je ploča veličine $2 * n$ koju je potrebno popločiti pločicama dimenzija $2 * 1$. Pločice je moguće rotirati. *Neka polja su blokirana i na njih nije moguće postaviti pločicu.* Na koliko načina je moguće ostvariti popločavanje? Ako nije moguće popločiti ploču ispišite $-1$. Pošto broj načina može biti jako velik, ispišite rješenje modulo 1e9 + 7.
+
+Ovaj jedan dodatni uvijet znatno nam otežava rješavanje zadatka jer prethodna formula više ne pomaže.
+
+### Rješenje
+
+Probajmo si olakšati zadatak s nekim opservacijama.
```

Comment:
Ovaj problem se može puno lakše i generičnije riješiti ako stanje proširiš na broj načina tako da su u stupcu i ispunjena
a) oba polja
b) samo gornje polje
c) samo donje polje
a u svim ostalim stupcima da je ispunjeno sve.
I onda slučajevi gdje ne možeš postaviti nešto mogu biti riješeni jednostavnije.

Reply:
Aha ok, ima smisla, to ću popravit onda

Reply:
Evo koda (primijeti oblik inputa)
```
#include <bits/stdc++.h>
using namespace std;

int main(){
    int n; cin>>n;

    vector<int> P(n);
    for (int i=0;i<n;i++){
        char c; cin>>c;
        if (c=='#'){
            P[i]|=0b01;
        }
    }
    for (int i=0;i<n;i++){
        char c; cin>>c;
        if (c=='#'){
            P[i]|=0b10;
        }
    }

    constexpr int MOD = 1000000007;


    vector<vector<int>> D(n, vector<int>(4));
    D[0][0b00] = (P[0]==0b00);
    D[0][0b01] = (P[0]==0b01);
    D[0][0b10] = (P[0]==0b10);
    D[0][0b11] = (P[0]==0b11) | (P[0]==0b00);

    for (int i=1;i<n;i++){
        if (P[i] == 0b00){
            D[i][0b00] = D[i-1][0b11];
            D[i][0b10] = D[i-1][0b01];
            D[i][0b01] = D[i-1][0b10];
            D[i][0b11] = (D[i-1][0b11] + D[i-1][0b00])%MOD;
        }
        if (P[i] == 0b01){
            D[i][0b01] = D[i-1][0b11];
            D[i][0b11] = D[i-1][0b01];
        }
        if (P[i] == 0b10){
            D[i][0b10] = D[i-1][0b11];
            D[i][0b11] = D[i-1][0b10];
        }
        if (P[i] == 0b11){
            D[i][0b11] = D[i-1][0b11];
        }
    }

    cout << D[n-1][0b11] << endl;
}
```